Add the Isle of Bute Discovery Centre Cinema

andrewferguson opened the pull request:
[This cinema](http://discoverycentrecinema.blogspot.com/) is a small independent (one screen, 97 seats) and the listings are on a Blogspot blog that is updated each month with a hand-edited HTML table.

The "lxml" package has been added to the archiver to support the correct parsing of XML Atom/RSS feeds in BeautifulSoup, such as those used by Blogspot.

andrewferguson committed:
Add the Isle of Bute Discovery Centre Cinema

This cinema is a small independent (one screen, 97 seats) and the
listings are on a Blogspot blog that is updated each month with a
hand-edited HTML table.

The "lxml" package has been added to the archiver to support the
correct parsing of XML Atom/RSS feeds in BeautifulSoup, such as
those used by Blogspot.

davidferguson commented on src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py:
```diff
@@ -0,0 +1,95 @@
+
+def parse_table_row(table_row: BeautifulSoup) -> [Showing]:
+    row_datas = [td.text for td in table_row.find_all('td')]
+    showings = [ extract_showing(row_datas, 0), extract_showing(row_datas, 1)]
```
```suggestion
    showings = [extract_showing(row_datas, 0), extract_showing(row_datas, 1)]
```
Remove the whitespace here please

davidferguson commented on src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py:
```diff
@@ -0,0 +1,95 @@
+        listings_table = html_soup.find('table')
+        for i, row in enumerate(listings_table.find_all('tr')):
+            if i == 0:
+                continue # skip the 'Film listings for <Month> row'
```
```suggestion
                continue  # skip the 'Film listings for <Month> row'
```
Two spaces before the comment please

davidferguson commented on src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py:
```diff
@@ -0,0 +1,95 @@
+def parse_table_row(table_row: BeautifulSoup) -> [Showing]:
+    row_datas = [td.text for td in table_row.find_all('td')]
+    showings = [ extract_showing(row_datas, 0), extract_showing(row_datas, 1)]
+    showings = [showing for showing in showings if showing != None]
```
```suggestion
    showings = [showing for showing in showings if showing is not None]
```

Should be `showing is not None`

davidferguson commented on src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py:
```diff
@@ -0,0 +1,95 @@
+    return showings
+
+
+def extract_showing(row_datas: [str], showing_number: int) -> Showing:
```
```suggestion
def extract_showing(row_datas: [str], showing_number: int) -> Optional[Showing]:
```

The type should be `Optional[Showing]`.

davidferguson commented on src/showingpreviously/cinemas/isle_of_bute_discovery_centre_cinema.py:
```diff
@@ -0,0 +1,95 @@
+import re, lxml, datetime
```
Please have one import per line. And `lxml` is not used in this file.

andrewferguson replied:
Changed to one import per line. And `lxml` is used for parsing the RSS feed using BeautifulSoup:
```
feed_soup = BeautifulSoup(feed_contents, features='xml')
```

davidferguson commented on setup.py:
```diff
@@ -23,5 +23,6 @@
         'requests',
         'beautifulsoup4',
         'selenium',
+        'lxml',
```
Where is `lxml` used?

andrewferguson replied:
`lxml` is used for parsing the RSS feed using BeautifulSoup:
```
feed_soup = BeautifulSoup(feed_contents, features='xml')
```

davidferguson reviewed (commented):
A few minor formatting changes, but otherwise looks good.

andrewferguson committed:
Fix formatting issues

andrewferguson committed:
Fix type error

andrewferguson committed:
Relax rules for what counts as an invalid showing

Previously a row in the table was considered invalid only if both
the title and time were blank. This is unsuitable as sometimes
the title field is used to display a message about why no showing
is occuring.

Changing the rules to result in an invalid showing if either the
title or time fields are blank results in the correct behaviour.

andrewferguson committed:
Allow for advance month publications

Sometimes two months' worth of listings will be published at once,
as is the case of November and December 2022. The problem with this
was that the cinema archiver was configured to only consider the
first XML feed item, and so would miss the other month.

This has been re-configured to accept all the showings in the
feed that are in the future.

andrewferguson commented:
The issues highlighted have been addressed, and a bug caused by the hand-edited HTML of this cinema's listings has been corrected.

It runs correctly, and is now ready for review.

davidferguson committed:
Merge branch 'master' into addcinema-isleofbutediscovery
